### PR TITLE
RFC: rails updates

### PR DESCRIPTION
**Pull request by tgaff**
The main goal here is modernizing and getting onboard with more recent trends in the industry

In addition, I've also added some thoughts on:

* services
* Stimulus and Turbo
* removed cucumber

Let me know what you think.  If some of this needs a lot more discussion I can pull it out and open a separate PR & chat/meeting for it.  

Things I'd like comments on:

* [x] Removing cucumber OK?  I think regular RSpec integration tests are much more common and usually easier to maintain these days. 
* [ ] Services?  I use these a lot - but maybe its not something we always want to recommend.
* JS
  * [ ] Should we get rid of most/all of unobtrusive JS?  I can't recall anyone requesting it recently.  I could convert that section into something more like "Rails JS" and leave UJS as a smaller topic within.
  * [x] I think its maybe out of place but should we talk about when to choose Hotwire vs. React/Vue vs. something else? (sounds like a blog post)
  * [ ] Should we keep out of the Hotwire stuff entirely for now?
* [x] Grammar / sentence structure / be my Grammarly please

### CODE REVIEW

File: languages/ruby/ruby_on_rails.md [hunk 1]
```diff
@@ -3,34 +3,38 @@
...
+There are many things that we will need to communicate and share with others developers during and after we’re done.  Please document the basics that help folks know how to install, configure and deploy the app.  We are committed to updating these documents when things change to keep the docs useful.  Given that that’s always going to be an ongoing task, please use documentation wisely.  Too much and it’s unusable and ends up out of date.  If a dragon hides in the system, write about it.  This starts with the repo's README.md but doesn't end there.  In some organizations a wiki may also be appropriate.  The integration tests especially can serve as a
```
Comment (mjelks) [inline languages/ruby/ruby_on_rails.md:10]: Assume still working on the last sentence? 

Also, I know with documentation we were even more explicit with certain items and used a `doc` or `readme` folder? 

e.g. `/app_name/readmes`

Comment (mjelks) [reply languages/ruby/ruby_on_rails.md:10]: `For more complex functionality and/or code that might touch external functionality, it might be recommended to extend this approach by creating a separate set of README(s) as needed. Ideally, this would be referenced and linked from the base README.md`

File: languages/ruby/ruby_on_rails.md [hunk 2]
```diff
@@ -3,34 +3,38 @@
...
 
 ## READMEs
-Please delete the Rails boilerplate one when starting a new app and start plumbing it with the what’s-where basics.  This page is what’s shown on the Github homepage for the project and is a good launching point to other pages on the wiki. Use Markdown when appropriate to improve formatting.
+Please delete the Rails boilerplate one when starting a new app and start plumbing it with the what’s-where basics.  This page is what’s shown on the Github homepage for the project and is a good launching point to other pages on the wiki. Use Markdown when appropriate to improve formatting.  Always aim for a balance between "enough to handle common tasks" and "too long to effectively find what I need".
 
-## Github Wiki
-Anything beyond the über-basics should be documented on the Github wiki pages.  Some examples of this might be “Deployment Anatomy,” “Search Architecture,” “Importing + Refreshing Application Content,” etc…
+## Wiki's
```
Comment (mjelks) [inline languages/ruby/ruby_on_rails.md:15]: I believe possessive is wrong in this case 

**Wikis**

Comment (mjelks) [reply languages/ruby/ruby_on_rails.md:15]: FYI here is the [GitHub link on the subject](https://docs.github.com/en/communities/documenting-your-project-with-wikis/about-wikis)

File: languages/ruby/ruby_on_rails.md [hunk 3]
```diff
@@ -3,34 +3,38 @@
...
+Generally these are to be avoided only because they tend to be out of date if used too heavily.  You should favor using **descriptive variable and method names**.  
 Rather, please use inline code comments when:
 * The following lines of code might use some syntax-fu (ie. anything overly-clever)
 * You’re working around a bug in a 3rd party library/api/model that we can’t control, i.e. when we have to write something that seems bad in order to use an external/immutable dependency.
 * Magic numbers: Explain your units and why the number exists.
-* Class documentation: Consider documenting the controller or model to give the big picture so that folks can find their bearings.
+* Class documentation: Consider documenting service and library classes.  They often can interact together in ways that might not be immediately obvious from just the names.  Model classes that are especially meaty can also benefit from documentation, especially as new team members come on-board.
 * In the Gemfile when locking a gem to a specific version: Please document why we must use that version. This makes it easier to know when it is safe/reasonable to upgrade the gem.
```
Comment (mjelks) [inline languages/ruby/ruby_on_rails.md:25]: I guess this ties in with above comment on more in-depth `readmes`

File: languages/ruby/ruby_on_rails.md [hunk 4]
```diff
@@ -106,20 +121,18 @@ extract value objects (and place in app/models)
...
+introduce view objects (helpers, ViewComponents and Decorators are all good directions to think in)
 extract decorators (consider using the draper gem)
+extract policy objects (there are a number of good gems for this)
 
 # Security
 
-As a starting point, please read http://guides.rubyonrails.org/security.html and become familiar with the concepts.  We must adhere to all of the points mentioned there.
-
-Another useful, quick read is http://blog.codeclimate.com/blog/2013/03/27/rails-insecure-defaults/
+As a starting point, please read https://guides.rubyonrails.org/security.html and become familiar with the concepts.  We must adhere to all of the points mentioned there.
```
Comment (mjelks) [inline languages/ruby/ruby_on_rails.md:130]: Do we want to mention response to Dependabot alerts in this section? i.e. what is the expected mitigation approach to brakeman and/or dependabot alerts. 

- Low
- Medium
- High
- Critical

Comment (mjelks) [reply languages/ruby/ruby_on_rails.md:130]: Or maybe a separate section after **Brakeman** to go over automated code security threat response.

File: languages/ruby/ruby_on_rails.md [hunk 5]
```diff
@@ -106,20 +121,18 @@ extract value objects (and place in app/models)
...
 
 ## Brakeman
 We include brakeman in our CI runs on both PRBuilder jobs and on commit-level jobs on develop when we can't use CodeClimate.
 
-The brakeman Jenkins job should be configured to fail if any new security vulnerabilities are found.
+The brakeman CI job should be configured to fail if any new security vulnerabilities are found.
 
 Remember to keep this gem updated to catch the latest list of vulnerabilities.
```
Comment (mjelks) [inline languages/ruby/ruby_on_rails.md:138]: Do we also want to explicitly call out **Routes** as a new section for this guide? 

This might also dovetail with an explicit **API** section if so desired as it seems a lot of our latest projects have had a heavy API component attached as well.

I know for a fact that we have all been snakebit every once in a while on proper routing techniques. I know it's in the Rails guide, but might be nice to explicitly call out best-practices and confirm them here. Perfectly ok to leave it out for this draft, however.

**Review by mjelks — COMMENTED** [review]
Overall this looks really good. I think we definitely have enough to warrant a (much-needed) update. 

Only real blocker atm seems to be dangling sentence? Assuming this is in a pseudo-WIP state.

File: languages/ruby/ruby_on_rails.md (hunk 2 above)
Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:15]: good catch

File: languages/ruby/ruby_on_rails.md (hunk 1 above)
Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:10]: > Assume still working on the last sentence?
Thanks for reading so thoroughly! :-) I had moved part of that around and apparently lost some along the way.

How does that line up with wikis then?  Should we have all 3 of README.md, readmes/my_problem.md and wiki/my_problem ?

> Also, I know with documentation we were even more explicit with certain items and used a doc or readme folder?

I kinda think wiki and a docs directory should be either/or for code.  Though that doesn't mean a wiki might not be used for other institutional knowledge.
Between the two, I personally prefer the readmes/ directory myself because its easier to update while coding, which in turn means its slightly more likely to be up to date.

Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:10]: @labzero/lz-engineering  If we have some level of consensus on a docs/ readmes/ directory I'm happy to add that as well.

File: languages/ruby/ruby_on_rails.md (hunk 4 above)
Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:132]: I think in many cases it might depend on the organization and the app.  There is a brakeman section below and I think it's OK as a overview.

File: languages/ruby/ruby_on_rails.md (hunk 5 above)
Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:140]: Rails routing can be a tricky thing.  It kinda has too many options.  I'm not sure what we'd say here other than maybe:

* use resourceful routes (when possible)
* don't overly nest

But the rails guide says both of those already so I'm not sure it deserves its own section.

File: languages/ruby/ruby_on_rails.md (hunk 3 above)
Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:27]: Actually... not necessarily.  The Class documentation being discussed here is in code descriptions.  There's actually a rubocop rule for this **and its enabled by default**.  
https://docs.rubocop.org/rubocop/cops_style.html#styledocumentation

The rubocop.yml in this repo is very out of date so I've been working on a new one.  (separate PR sometime soon)  A *lot* of rules have been added since then and many of them are now enabled by default.  My config for that rule currently looks like:

```rb
Style/Documentation: # MVC & Rails conventions are *often* sufficient documentation for these classes
  Exclude:
    - app/mailers/**/*
    - app/models/**/*
    - app/components/**/*
    - app/controllers/**/*
    - app/helpers/**/*
    - config/**/*
    - db/**/*
```

Would love to hear if others disagree with this approach or not.  I've left several directories in the "include" list intentionally: `services`, `lib`.  I'd like to have a way to include `app/controllers/concerns` since it may be a bit less obvious from just seeing routing and controller methods.

File: languages/ruby/ruby_on_rails.md [hunk 6]
```diff
@@ -101,25 +125,25 @@ This is a fancy way of saying that your views and controller code is not the pla
...
+- extract form objects (and place in app/forms)
+- extract query objects (and place in app/models)
+- introduce view objects (helpers, ViewComponents and Decorators are all good directions to think in)
+- extract decorators (consider using the draper gem)
+- extract policy objects (there are a number of good gems for this)
 
-As a starting point, please read http://guides.rubyonrails.org/security.html and become familiar with the concepts.  We must adhere to all of the points mentioned there.
+# Security
 
-Another useful, quick read is http://blog.codeclimate.com/blog/2013/03/27/rails-insecure-defaults/
```
Comment (tgaff) [inline languages/ruby/ruby_on_rails.md:117]: I had removed this as it was a 404 - but I've found a new link for it.  That new link has been updated but its maybe less useful now than it used to be.  https://codeclimate.com/blog/rails-insecure-defaults

Who thinks we should include it? ✋

File: languages/ruby/ruby_on_rails.md (hunk 3 above)
Comment (mjelks) [reply languages/ruby/ruby_on_rails.md:31]: I like the idea of buffing out the rubocop / code styling as a separate thing. It definitely warrants proper attention.

File: languages/ruby/ruby_on_rails.md (hunk 4 above)
Comment (mjelks) [reply languages/ruby/ruby_on_rails.md]: great. lots of these comments are stream-of-consciousness. agreed we don't want this to attempt to be all-encompassing.

File: languages/ruby/ruby_on_rails.md (hunk 6 above)
Comment (mjelks) [reply languages/ruby/ruby_on_rails.md:117]: I'm not in favor of a URL with 2013 in it :P maybe there's a newer/fresher one out there, or prob best to avoid as many external URLs as possible.

Comment (tgaff) [reply languages/ruby/ruby_on_rails.md:117]: Yes, the one I posted above is the updated version (looks like they re-did all their blog URLs to remove the dates).

I think its likely the same article, but it was updated in March.  That being said.... I'm not totally sure its as relevant today as it was back when this was originally added.  They've marked most of their original points as fixed and there aren't as many remaining.  Of those I think that "yes" we should be aware of, but some of them are a stretch and they're less "by default" now.  

I'd argue at this point that a rails setup has more and better built-in security than many of the other options out there (e.g. express) and that in my opinion is one of the selling points.  That's not to say you don't have to be smart and know the framework though!

File: languages/ruby/ruby_on_rails.md [hunk 7]
```diff
@@ -3,34 +3,44 @@
...
+
+Furthermore, consider the names and structure of your integration tests. Well named tests can act as a form of documentation, providing insights into the systems features, shedding light on our original design intentions, and highlighting edge cases.
 
 ## READMEs
-Please delete the Rails boilerplate one when starting a new app and start plumbing it with the what’s-where basics.  This page is what’s shown on the Github homepage for the project and is a good launching point to other pages on the wiki. Use Markdown when appropriate to improve formatting.
+Please delete the Rails boilerplate one when starting a new app and start plumbing it with the what’s-where basics.  This page is what’s shown on the Github homepage for the project and is a good launching point to other pages in the documentation. Use Markdown when appropriate to improve formatting.  Always aim for a balance between "enough to handle common tasks" and "too long to effectively find what I need".
```
Comment (Shiun) [inline languages/ruby/ruby_on_rails.md:17]: Should we be more prescriptive?

At the very least, the page should include project purpose and setup.

**Comment by Shiun** [comment]
I think this looks good.  TBH, Kompliant was my first exposure to Turbo/Stimulus so I don't have a strong opinion re: React vs Turbo/Stimulus.  That said, I definitely found it to be a good alternative to React for a responsive application.  And the guidelines you mentioned are good recommendations (re: general purpose Stimulus controllers that can be used in multiple interactive UI components).  

I would defer to @JeffreyATW for his opinion on unobtrusive JS.

**Comment by tgaff** [comment]
Thanks for reviewing Shiun

> I think this looks good. TBH, Kompliant was my first exposure to Turbo/Stimulus so I don't have a strong opinion re: React vs Turbo/Stimulus. 

I don't want to prescribe one over the other as a rule.  They're both going to be around and both are reasonable, valid and officially supported via `rails new ....` commands.  I'll re-read and try to clarify that there's a choice here and it depends on the type of application you're building and its goals.

File: languages/ruby/ruby_on_rails.md (hunk 7 above)
Comment (tgaff) [reply languages/ruby/ruby_on_rails.md]: resolved?

**Comment by tgaff** [comment]
Updates:

* re-wrote JS section
  * added import maps vs. bundling decision
  * reduced UJS section content and tied it to Hotwire section
  * reduced Hotwire section content
  * linked to Rails guides
* added more on README per Shiun

**Comment by mjelks** [comment]
Based on the latest updates, I believe you've addressed most if not all of your checklist? I am going to approve and say this is definitely an improvement over the existing guide. 

The compromise of UJS + Hotwire documentation means we now have both listed in the README.

**Review by mjelks — APPROVED** [review]
Based on the latest updates, I believe you've addressed most if not all of your checklist? I am going to approve and say this is definitely an improvement over the existing guide.

The compromise of UJS + Hotwire documentation means we now have both listed in the README.

**Comment by tgaff** [comment]
> The compromise of UJS + Hotwire documentation means we now have both listed in the README.

Yep.  I could delete them both....?

Thanks for reviewing!

**Comment by tgaff** [comment]
@Shiun - leaving this a little longer in case you want to review again.  Will likely merge early next week.

File: languages/ruby/ruby_on_rails.md [hunk 8]
```diff
@@ -3,34 +3,47 @@
...
 
 ## READMEs
-Please delete the Rails boilerplate one when starting a new app and start plumbing it with the what’s-where basics.  This page is what’s shown on the Github homepage for the project and is a good launching point to other pages on the wiki. Use Markdown when appropriate to improve formatting.
 
-## Github Wiki
-Anything beyond the über-basics should be documented on the Github wiki pages.  Some examples of this might be “Deployment Anatomy,” “Search Architecture,” “Importing + Refreshing Application Content,” etc…
+Please delete the Rails boilerplate one when starting a new app and start plumbing it with the what’s-where basics.  This page is what’s shown on the Github homepage for the project and is a good launching point to other pages in the documentation. Use Markdown when appropriate to improve formatting.  
+
+Always aim for a balance between "enough to handle common tasks" and "too long to effectively find what I need". At the very least, the page should include project setup and any description necessary to find the right repo.
```
Comment (Shiun) [inline languages/ruby/ruby_on_rails.md:20]: This looks good.  Thanks for adding this.

**Review by Shiun — APPROVED** [review]
This is a good update to our guide.
